frontend/apps/reader/modules/readertypography.lua

```diff
@@ -0,0 +1,802 @@
+local BD = require("ui/bidi")
+local Device = require("device")
+local Event = require("ui/event")
+local InfoMessage = require("ui/widget/infomessage")
+local InputContainer = require("ui/widget/container/inputcontainer")
+local MultiConfirmBox = require("ui/widget/multiconfirmbox")
+local UIManager = require("ui/uimanager")
+local logger = require("logger")
+local util = require("util")
+local _ = require("gettext")
+local C_ = _.pgettext
+local T = require("ffi/util").template
+local Screen = Device.screen
+
+-- Mostly for migrating hyph settings, and to know the dict
+-- left and right hyph min values (2/2 when not specified)
+local HYPH_DICT_NAME_TO_LANG_TAG = {
+    ["@none"]                = { "en" },
+    ["@softhyphens"]         = { "en" },
+    ["@algorithm"]           = { "en" },
+    ["Bulgarian.pattern"]    = { "bg" },
+    ["Catalan.pattern"]      = { "ca" },
+    ["Czech.pattern"]        = { "cs" },
+    ["Danish.pattern"]       = { "da" },
+    ["Dutch.pattern"]        = { "nl" },
+    ["English_GB.pattern"]   = { "en-GB" },
+    ["English_US.pattern"]   = { "en-US" },
+    ["Finnish.pattern"]      = { "fi" },
+    ["French.pattern"]       = { "fr", 2, 1 },
+    ["Galician.pattern"]     = { "gl" },
+    ["German.pattern"]       = { "de" },
+    ["Greek.pattern"]        = { "el" },
+    ["Hungarian.pattern"]    = { "hu" },
+    ["Icelandic.pattern"]    = { "is" },
+    ["Irish.pattern"]        = { "ga" },
+    ["Italian.pattern"]      = { "it" },
+    ["Norwegian.pattern"]    = { "no" },
+    ["Polish.pattern"]       = { "pl" },
+    ["Portuguese.pattern"]   = { "pt" },
+    ["Roman.pattern"]        = { "ro" },
+    ["Russian_EnGB.pattern"] = { "ru-GB" },
+    ["Russian_EnUS.pattern"] = { "ru-US" },
+    ["Russian.pattern"]      = { "ru" },
+    ["Slovak.pattern"]       = { "sk" },
+    ["Slovenian.pattern"]    = { "sl" },
+    ["Spanish.pattern"]      = { "es" },
+    ["Swedish.pattern"]      = { "sv" },
+    ["Turkish.pattern"]      = { "tr" },
+    ["Ukrain.pattern"]       = { "uk" },
+}
+
+-- Languages to be shown in the menu.
+-- Aliases are language tags that could possibly be found in the book
+-- language metadata, that we wish to map to the correct lang tag.
+-- Features:
+--   H = language specific hyphenation dictionary
+--   b = language specific line breaking rules
+--   B = language specific additional line breaking tweaks
+-- Update them when language tweaks and features are added to crengine/src/textlang.cpp
+local LANGUAGES = {
+    -- lang-tag          aliases    features    menu title
```

Comment:
Random fun fact: 2 letters language codes are legacy ISO 639-1 (-ish, I think the language_DIALECT bastardization is a GNU thing) & ISO 639-2 (where I learned that fre is for French Belgian ;)).

Reply:
Or it's actually https://tools.ietf.org/html/rfc5646, as `ReaderTypography:parseLanguageTag()` mentions below ;).

Reply:
Well, I did not dig into that much (or at all). Not much interested in learning about that - so, if anyone wants to specialize and add stuff so we do the right thing with tags we'll meet in the wild, please jump on it :) (here and in textlang.cpp).
One thing I get is that Harfbuzz has lots of code to handle that/translate, and I hope it will do the right thing with what we provide as is. (Except may be with the 3 letters language code like "fre" that we translate here...)

Reply:
One small comment: it should be “Ukrainian” on line 49

Reply:
I do have at least one Belgian French book that uses “ ”. I don't know if that's specifically Belgian and if it actually makes any difference to the algorithm.

Reply:
> I do have at least one Belgian French book that uses “ ”

Hard to recognize in github:
![image](https://user-images.githubusercontent.com/24273478/79694815-14cc3a80-8273-11ea-8c37-a5d41979f1e7.png)

but pasting them into google, it sees them as left and right double quotation mark, so perfectly balanced and handled by:
https://github.com/koreader/crengine/blob/65d4adf843b4c85627187d4e2f685b8f42dfa1f4/crengine/src/textlang.cpp#L395-L398

So, whether there are space inside or not, they should stick to their first and last content.

Reply:
> One small comment: it should be “Ukrainian” on line 49

That's just a filename (well, that's how that file is named for ages), that might be saved in current users settings.
I was going to say it won't be shown, but it will, as the selected hyph dict.
So, dunno if I should rename it (may be later, when things are migrated), or if we should have another table for translating typos in filenames...

Reply:
@poire-z
Sorry I didn’t know that :) Looks like it comes all the way from the first ones who PR’ed Ukrainian language files

Reply:
@mergen3107 On the contrary, thanks for pointing it out!

@poire-z If I go to hyphenation it already says Ukrainian, which probably implies we already have that?

Reply:
![image](https://user-images.githubusercontent.com/24273478/79695403-fec07900-8276-11ea-8096-4d8da2a50ac8.png)
We have the other mapping lang_tag => language shows alongside typography, but we can't really use it, as zh-CN would show "Chinese (Simplified)" while it uses English_US.pattern.

I don't mind a small typo table or better: a function to remove the .pattern and fix some known typos, to be used instead of:
`hyph_dict_name = hyph_dict_name:gsub(".pattern$", "")`

Reply:
Well, I guess I'll update the HYPH_DICT_NAME_TO_LANG_TAG table to include a translatable string of the language - instead of using the filename stripped from .pattern, which won't be translatable.
If we remove the .pattern for pretty menus, I guess there's no more reason to keep the relation to the hyph dict filename.